src/MinimalDelegation.sol

```diff
@@ -69,52 +75,42 @@ contract MinimalDelegation is IERC7821, ERC1271, EIP712, ERC4337Account, Receive
     function executeUserOp(PackedUserOperation calldata userOp, bytes32) external onlyEntryPoint {
         // Parse the keyHash from the signature. This is the keyHash that has been pre-validated as the correct signer over the UserOp data
         // and must be used to check further on-chain permissions over the call execution.
-        // TODO: Handle keyHash authorization.
-        // (bytes32 keyHash,) = userOp.signature.unwrap();
+        
+        (bytes32 keyHash,) = userOp.signature.unwrap();
 
         // The mode is only passed in to signify the EXEC_TYPE of the calls.
         (bytes32 mode, bytes calldata executionData) = userOp.callData.removeSelector().decodeBytes32Bytes();
         if (!mode.isBatchedCall()) revert IERC7821.UnsupportedExecutionMode();
         Call[] calldata calls = executionData.decodeCalls();
 
-        _dispatch(mode, calls);
-    }
-
-    /// @dev The mode is passed to allow other modes to specify different types of opData decoding.
-    function _authorizeOpData(bytes32, Call[] calldata calls, bytes calldata opData) private {
-        (uint256 nonce, bytes calldata wrappedSignature) = opData.decodeUint256Bytes();
-        _useNonce(nonce);
-
-        bytes32 digest = _hashTypedData(calls.toSignedCalls(nonce).hash());
-
-        (bytes32 keyHash, bytes calldata signature) = wrappedSignature.unwrap();
-        Key memory key = _getKey(keyHash);
-
-        IHook hook = keySettings[keyHash].hook();
-
-        /// TODO: Handle key expiry check.
-        bool isValid = hook.hasPermission(HooksLib.VERIFY_SIGNATURE_FLAG)
-            ? hook.verifySignature(keyHash, digest, signature)
-            : key.verify(digest, signature);
-
-        if (!isValid) revert IERC7821.InvalidSignature();
+        _dispatch(mode, calls, keyHash);
     }
 
     /// @dev Dispatches a batch of calls.
-    function _dispatch(bytes32 mode, Call[] calldata calls) private {
+    function _dispatch(bytes32 mode, Call[] calldata calls, bytes32 keyHash) private {
         bool shouldRevert = mode.shouldRevert();
 
         for (uint256 i = 0; i < calls.length; i++) {
-            (bool success, bytes memory output) = _execute(calls[i]);
+            (bool success, bytes memory output) = _execute(calls[i], keyHash);
             // Reverts with the first call that is unsuccessful if the EXEC_TYPE is set to force a revert.
             if (!success && shouldRevert) revert IERC7821.CallFailed(output);
         }
     }
 
-    // Execute a single call.
-    function _execute(Call calldata _call) private returns (bool success, bytes memory output) {
+    /// @dev Executes a low level call using execution hooks if set
+    function _execute(Call calldata _call, bytes32 keyHash) internal returns (bool success, bytes memory output) {
+        // Per ERC7821, replace address(0) with address(this)
         address to = _call.to == address(0) ? address(this) : _call.to;
+
```

Comment:
We still need to check the admin settings on the key here, maybe a todo comment?

Reply:
good callout